### PR TITLE
[BEAM-2713] Support Python native types in Beam typehints

### PR DESCRIPTION
**Pull request by chuanyu**
Add support for Python native types in Beam typehints.

### CODE REVIEW

**Comment by chuanyu** [comment]
CC @robertwb

**Comment by aaltay** [comment]
cc: @aaltay

Hey @chuanyu, thank you for your first contribution. I believe `typing` module is only available in python 3.5+, Beam Python SDK currently does not support Python 3. And this change should happen together with a grand change that upgrades all of SDK to support Python 3. Would you be interested in doing that? If you are interested, a good step would be preparing a plan and sharing with the dev list.

**Comment by asfgit** [comment]
FAILURE
 
--none--

**Comment by chuanyu** [comment]
Hi @aaltay, thanks for the quick reply! The `typing` module has actually been [backported to pre-3.5 versions](https://pypi.python.org/pypi/typing/3.6.1), and type annotations are even available in Python 2.7 and up as type comments. It would be useful to be able to use the `typing` module types for Beam typehints as well, especially when the types are complicated. 

No worries if you'd prefer not to include this change in isolation - perhaps use of typing in pre-3.5 is not common enough for this to be a frequent request.

**Comment by charlesccychen** [comment]
CC: @robertwb

Perhaps we should consider encouraging using standard Python typing instead of our type annotations since @chuanyu indicates it is broadly available.  Robert, have you seen any issues that preclude this use?

**Comment by robertwb** [comment]
We didn't use the typing module at the time because it was not yet backported; if it's sufficiently expressive (probably) we should migrate to it. This PR would be a good start for backwards compatibility.

**Comment by charlesccychen** [comment]
R: @charlesccychen

File: sdks/python/apache_beam/typehints/native_type_compatibility.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def convert_to_beam_type(typ):
+  """Convert a given typing type to a Beam type.
+
+  Args:
+    typ: typing type.
+
+  Returns:
+    The given type converted to a Beam type as far as we can do the converion.
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: "conversion"

File: sdks/python/apache_beam/typehints/native_type_compatibility.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+def _safe_issubclass(derived, parent):
+  """Like issubclass, but swallows TypeErrors.
+
+  This is useful for when either parameter might not actually be a class,
+  e.g. typing.Union isn't actually a class.
+
+  Args:
+    derived: As in issubclass
+    parent: As in issubclass
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: nit: period at end of sentence.

File: sdks/python/apache_beam/typehints/native_type_compatibility.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def _safe_issubclass(derived, parent):
+  """Like issubclass, but swallows TypeErrors.
+
+  This is useful for when either parameter might not actually be a class,
+  e.g. typing.Union isn't actually a class.
+
+  Args:
+    derived: As in issubclass
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: nit: period at end of sentence.

File: sdks/python/apache_beam/typehints/native_type_compatibility.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Module to convert Python's native typing types to Beam types."""
+
+try:
+  import typing
+except ImportError:
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: Please remove conditional import and add the dependency to setup.py, as suggested, pinned to the minor version >=3.6.0, <3.7.0.

File: sdks/python/apache_beam/typehints/typed_pipeline_test.py [hunk 5]
```diff
@@ -98,6 +98,38 @@ def process(self, element):
       [1, 2, 3] | (beam.ParDo(my_do_fn) | 'again' >> beam.ParDo(my_do_fn))
 
 
+# NativeTypesTest is only run if the typing module is available.
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/typed_pipeline_test.py]: Please remove conditional import, as we will depend directly on `typing`.

File: sdks/python/apache_beam/typehints/native_type_compatibility.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    # Nullary types (e.g. Any) don't accept empty tuples as arguments.
+    return matched_entry.beam_type
+  elif arity == 1:
+    # Unary types (e.g. Optional) don't accept 1-tuples as arguments
+    return matched_entry.beam_type[typs[0]]
+  else:
+    return matched_entry.beam_type[tuple(typs)]
+
+
+def convert_to_beam_types(args):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: This naming is somewhat confusing given the existence of `convert_to_beam_type` in the same module, but I can't come up with a more reasonable name.  (no action needed, unless you can come up with something better)

File: sdks/python/apache_beam/typehints/native_type_compatibility_test.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+# limitations under the License.
+#
+
+"""Test for Beam type compatibility library."""
+
+import sys
+
+try:
+  import typing
+except ImportError:
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility_test.py]: Please remove this guard.

**Comment by chuanyu** [comment]
Thanks for the review, @charlesccychen ! Addressed your comments. PTAL.

File: sdks/python/apache_beam/typehints/typehints.py [hunk 8]
```diff
@@ -507,7 +507,7 @@ def __getitem__(self, py_type):
       raise TypeError('An Option type-hint only accepts a single type '
                       'parameter.')
 
-    return Union[py_type, None]
+    return Union[py_type, type(None)]
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/typehints.py]: This looks fine to me, but I want to check with @robertwb whether there is a subtlety I may be missing here.

File: sdks/python/apache_beam/typehints/native_type_compatibility.py (hunk 2 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: > **charlesccychen** wrote:
> nit: period at end of sentence.


Done.

File: sdks/python/apache_beam/typehints/native_type_compatibility.py (hunk 3 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: > **charlesccychen** wrote:
> nit: period at end of sentence.


Done.

File: sdks/python/apache_beam/typehints/native_type_compatibility.py (hunk 1 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: > **charlesccychen** wrote:
> "conversion"


Done.

File: sdks/python/apache_beam/typehints/native_type_compatibility.py (hunk 4 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py]: > **charlesccychen** wrote:
> Please remove conditional import and add the dependency to setup.py, as suggested, pinned to the minor version >=3.6.0, <3.7.0.


Done.

File: sdks/python/apache_beam/typehints/native_type_compatibility_test.py (hunk 7 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility_test.py]: > **charlesccychen** wrote:
> Please remove this guard.


Please fix.

**Comment by chuanyu** [comment]
Removed import guard. PTAL!

File: sdks/python/apache_beam/typehints/native_type_compatibility_test.py (hunk 7 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/native_type_compatibility_test.py]: > **charlesccychen** wrote:
> Please fix.


Done.

File: sdks/python/apache_beam/typehints/typed_pipeline_test.py (hunk 5 above)
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/typed_pipeline_test.py]: > **charlesccychen** wrote:
> Please remove conditional import, as we will depend directly on `typing`.


Done.

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Pull request closed without merging**